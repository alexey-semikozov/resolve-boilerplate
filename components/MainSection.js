import React from 'react';
import PropTypes from 'prop-types';
import TodoItem from './TodoItem';
import Footer from './Footer';

export const SHOW_ALL = 'all';
export const SHOW_COMPLETED = 'completed';
export const SHOW_ACTIVE = 'active';

const TODO_FILTERS = {
  [SHOW_ALL]: () => true,
  [SHOW_ACTIVE]: todo => !todo.completed,
  [SHOW_COMPLETED]: todo => todo.completed
};

function renderFooter(completedCount, { todos, filter, actions }) {
  const activeCount = todos.length - completedCount;
  const titles = {
    [SHOW_ALL]: 'All',
    [SHOW_ACTIVE]: 'Active',
    [SHOW_COMPLETED]: 'Completed'
  };

  if (todos.length) {
    return (
      <Footer
        completedCount={completedCount}
        activeCount={activeCount}
        filter={filter}
        titles={titles}
      />
    );
  }
}

function MainSection(props) {
  const { todos, actions, filter } = props;
  const filteredTodos = todos.filter(TODO_FILTERS[filter]);
  const completedCount = todos.reduce(
    (count, todo) => (todo.completed ? count + 1 : count),
    0
  );

  return (
    <section className="main">
      <ul className="todo-list">
        {filteredTodos.map(todo => (
          <TodoItem key={todo.aggregateId} todo={todo} {...actions} />
        ))}
      </ul>
      {renderFooter(completedCount, props)}
    </section>
  );
}

MainSection.propTypes = {
  todos: PropTypes.array.isRequired,
  filter: PropTypes.string.isRequired,
  actions: PropTypes.object.isRequired
};

export default MainSection;
