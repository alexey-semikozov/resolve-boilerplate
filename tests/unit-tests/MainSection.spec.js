import React from 'react';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import renderer from 'react-test-renderer';
import MainSection from '../../components/MainSection';

const setup = propOverrides => {
  const props = Object.assign(
    {
      todos: [
        {
          text: 'Use Redux',
          completed: false,
          aggregateId: 0
        },
        {
          text: 'Run the tests',
          completed: true,
          aggregateId: 1
        }
      ],
      filter: 'all',
      actions: {
        editTodo: jest.fn(),
        deleteTodo: jest.fn(),
        completeTodo: jest.fn()
      }
    },
    propOverrides
  );

  const shallowrenderer = new ReactShallowRenderer();
  shallowrenderer.render(<MainSection {...props} />);
  const output = shallowrenderer.getRenderOutput();

  const component = renderer.create(<MainSection {...props} />);
  const tree = component.toJSON();

  return {
    props,
    output,
    tree
  };
};

describe('components', () => {
  describe('MainSection', () => {
    it('should render container', () => {
      const { tree } = setup();
      expect(tree).toMatchSnapshot();
    });
  });
});
