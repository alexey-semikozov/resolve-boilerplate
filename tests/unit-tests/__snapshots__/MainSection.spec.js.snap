// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components MainSection should render container 1`] = `
<section
  className="main"
>
  <ul
    className="todo-list"
  >
    <li
      className=""
    >
      <div
        className="view"
      >
        <input
          checked={false}
          className="toggle"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          onDoubleClick={[Function]}
        >
          Use Redux
        </label>
        <button
          className="destroy"
          onClick={[Function]}
        />
      </div>
    </li>
    <li
      className="completed"
    >
      <div
        className="view"
      >
        <input
          checked={true}
          className="toggle"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          onDoubleClick={[Function]}
        >
          Run the tests
        </label>
        <button
          className="destroy"
          onClick={[Function]}
        />
      </div>
    </li>
  </ul>
  <footer
    className="footer"
  >
    <span
      className="todo-count"
    >
      <strong>
        1
      </strong>
       
      item
       left
    </span>
    <ul
      className="filters"
    >
      <li>
        <a
          className="selected"
          href="/all"
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        >
          All
        </a>
      </li>
      <li>
        <a
          className=""
          href="/active"
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        >
          Active
        </a>
      </li>
      <li>
        <a
          className=""
          href="/completed"
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
            }
          }
        >
          Completed
        </a>
      </li>
    </ul>
  </footer>
</section>
`;
